### DISCUSSION TITLE
Preview feature feedback: `relationJoins`

### DISCUSSION THREAD

**Discussion by janpio** [prisma/prisma#22288] [Feedback]
Support for the **relation queries using database native `JOIN`s** is in preview behind the `relationJoins` preview feature. 

- Documentation: https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#relation-load-strategies-preview
- Launch blog post: https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview

Please share your feedback about `relationJoins` released in [v5.7.0 for PostgreSQL](https://github.com/prisma/prisma/releases/tag/5.7.0) and [v5.10.0 for MySQL](https://github.com/prisma/prisma/releases/tag/5.10.0) in this discussion.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repository.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this discussion.
- Need help or have questions? Join the [Prisma Discord](https://pris.ly/discord) community and ask away! We're here to support you.

**GA planned for March-August 2025**

**Comment by Scalahansolo** [comment]
Hey Jan... Just wanted to post some initial feedback. Given the 1-1 example query yall showed off I ran that example against our data and it looks like below...

```
SELECT
	"t1"."id",
	"t1"."name",
	"Task_Status"."__prisma_data__" AS "status"
FROM
	"public"."TeamTask" AS "t1"
	LEFT JOIN LATERAL (
		SELECT
			COALESCE(JSON_AGG("__prisma_data__"), '[]') AS "__prisma_data__"
		FROM (
			SELECT
				"t4"."__prisma_data__"
			FROM (
				SELECT
					JSON_BUILD_OBJECT('id', "t3"."id", 'name', "t3"."name", 'workspaceId', "t3"."workspaceId") AS "__prisma_data__"
				FROM (
					SELECT
						"t2".*
					FROM
						"public"."TaskStatus" AS "t2"
					WHERE
						"t1"."statusId" = "t2"."id") AS "t3") AS "t4") AS "t5") AS "Task_Status" ON TRUE
LIMIT 1000;
```

This compared to a simple left join like below...

```
SELECT
	*
FROM
	"TeamTask" tt
	LEFT JOIN "TaskStatus" ts ON tt. "statusId" = ts.id
LIMIT 1000;
```

The simple left join on this 1-1 relation is about 2-10x (depending on the limit) faster compared to what prisma would generate for this query.

It's also worth mentioning that the lower query is also pulling back significantly more data as it's doing a select star compared to a subset of the information in the first query.

**Reply by Weakky** [reply]
Hey @Scalahansolo, thanks for letting us know. Would you mind providing us with an `EXPLAIN ANALYZE` output of both queries (simple left join vs prisma generated query) on your dataset?

Thank you!

**Reply by Scalahansolo** [reply]
Id be happy to. Is there somewhere I can give y'all this information that isn't so public?

**Reply by janpio** [reply]
You can email me at jan@prisma.io and I will forward it to Flavian.

**Reply by janpio** [reply]
Can you also include the query you are running, and maybe also the Prisma schema that is based on? Then we should be able to figure that out.

**Reply by janpio** [reply]
Thanks, email received and we will take a look.

**Reply by syenchuk** [reply]
The correct way to do a SQL query to retrieve a user and their profile (from [the release page example](https://github.com/prisma/prisma/releases/tag/5.7.0)) is simply:

```sql
SELECT *
FROM user U
INNER JOIN profile P ON P.user_id = U.id
WHERE U.id = 1
```

Why do we end up with this instead?

```sql
SELECT
    "t1"."id",
    "t1"."name",
    "User_profile"."__prisma_data__" AS "profile"
FROM
    "public"."User" AS "t1"
    LEFT JOIN LATERAL (
        SELECT
            COALESCE(JSON_AGG("__prisma_data__"), '[]') AS "__prisma_data__"
        FROM
            (
                SELECT
                    "t4"."__prisma_data__"
                FROM
                    (
                        SELECT
                            JSON_BUILD_OBJECT(
                                'id',
                                "t3"."id",
                                'bio',
                                "t3"."bio",
                                'userId',
                                "t3"."userId"
                            ) AS "__prisma_data__"
                        FROM
                            (
                                SELECT
                                    "t2".*
                                FROM
                                    "public"."Profile" AS "t2"
                                WHERE
                                    "t1"."id" = "t2"."userId"
                            ) AS "t3"
                    ) AS "t4"
            ) AS "t5"
    ) AS "User_profile" ON TRUE
WHERE "t1"."id" = $1
LIMIT $2
```

I don't understand how anyone in their right mind could think of this as a "good idea".

**Reply by kimchinosys** [reply]
Any updates on this? Would love to use the new `join` feature with Prisma but these raw SQL queries generated like above are very slow.

**Reply by BracketJohn** [reply]
Checkout / upvote issue #22596 to track this, I think this is probably what you are experiencing (as do we, commented below)

**Reply by janpio** [reply]
Is your feedback @kimchinosys related to 1-1 relations as @Scalahansolo mentioned in his comment, or more general about all queries when using `relationJoins`? If it is _not_ just about 1-1 relations, please make sure to check if #22596 matches your case and either comment there with more concrete examples, or if not open a new issue and give us some more information. We need your help to understand the problem.

**Reply by kimchinosys** [reply]
That's the same issue for us, thank you for sharing the link. Will keep track of it.

**Comment by mhammerc** [comment]
Our server test suite passed with the `relationJoins` preview feature. Good job!

**Reply by kasir-barati** [reply]
JFYI: https://www.cybertec-postgresql.com/en/understanding-lateral-joins-in-postgresql/#comment-1366

It is not all that great if you have large data sets.

Any comment on this @janpio? I mean did you know about this while developing this new feature?

Just for the record, I really do not like how ORMs sends separate queries to the underlying database (though I read [here](https://github.com/prisma/prisma/discussions/25407#discussion-7300305) that Prisma perform some kind of optimization and batching to boost the performance but I was not able to find any clue in the Prisma's official doc). So I am really looking forward to this feature.

**Comment by ajdiyassin** [comment]
What are the benefits of using database-native JOINs? does this increase performance?

**Reply by janpio** [reply]
Yes, usually it does. If it does not, please let us know which queries are affected.

**Comment by samuelcole** [comment]
`where: { volunteerEventImage: { isNot: null } }` on an include seems to fail (removing that fixes the query)

```
Invalid `prisma.volunteer.findUnique()` invocation:

The table `(not available)` does not exist in the current database.
```

``` js
const volunteer = await prisma.volunteer.findUnique({
      where: {
        id: volunteerId,
      },
      select: {
        name: true,
        volunteerEvents: {
          select: { volunteerEventImage: true },
          where: { volunteerEventImage: { isNot: null } },
          orderBy: [
            { archivedAt: "desc" },
            { endTime: "desc" },
            { startTime: "desc" },
          ],
          take: 1,
        },
      },
    });
```

``` sql
SELECT
  "t1"."id", 
  "t1"."name", 
  "Volunteer_volunteerEvents_m2m"."__prisma_data__" AS "volunteerEvents" 
FROM 
  "public"."Volunteer" AS "t1" 
  LEFT JOIN LATERAL (
    SELECT 
      COALESCE(
        JSON_AGG("__prisma_data__"), 
        '[]'
      ) AS "__prisma_data__" 
    FROM 
      (
        SELECT 
          "Volunteer_volunteerEvents"."__prisma_data__" 
        FROM 
          "public"."_VolunteerToVolunteerEvent" AS "t2" 
          LEFT JOIN LATERAL (
            SELECT 
              JSON_BUILD_OBJECT(
                'volunteerEventImage', "VolunteerEvent_volunteerEventImage"."__prisma_data__"
              ) AS "__prisma_data__", 
              "t4"."archivedAt", 
              "t4"."endTime", 
              "t4"."startTime", 
              "t4"."id" 
            FROM 
              (
                SELECT 
                  "t3".* 
                FROM 
                  "public"."VolunteerEvent" AS "t3" 
                WHERE 
                  "t2"."B" = "t3"."id" 
                  /* root select */
                  ) AS "t4" 
              LEFT JOIN LATERAL (
                SELECT 
                  COALESCE(
                    JSON_AGG("__prisma_data__"), 
                    '[]'
                  ) AS "__prisma_data__" 
                FROM 
                  (
                    SELECT 
                      "t9"."__prisma_data__" 
                    FROM 
                      (
                        SELECT 
                          JSON_BUILD_OBJECT(
                            'id', "t8"."id", 'createdAt', "t8"."createdAt", 
                            'url', "t8"."url", 'volunteerEventId', 
                            "t8"."volunteerEventId"
                          ) AS "__prisma_data__" 
                        FROM 
                          (
                            SELECT 
                              "t7".* 
                            FROM 
                              "public"."VolunteerEventImage" AS "t7" 
                            WHERE 
                              "t4"."id" = "t7"."volunteerEventId" 
                              /* root select */
                              ) AS "t8" 
                          /* inner select */
                          ) AS "t9" 
                      /* middle select */
                      ) AS "t10" 
                  /* outer select */
                  ) AS "VolunteerEvent_volunteerEventImage" ON true
          ) AS "Volunteer_volunteerEvents" ON true 
          LEFT JOIN "public"."VolunteerEventImage" AS "j1" ON ("j1"."volunteerEventId") = ("public"."VolunteerEvent"."id") 
        WHERE 
          (
            "t2"."A" = "t1"."id" 
            AND (
              NOT ("j1"."volunteerEventId" IS NULL)
            )
          ) 
        ORDER BY 
          "Volunteer_volunteerEvents"."archivedAt" DESC, 
          "Volunteer_volunteerEvents"."endTime" DESC, 
          "Volunteer_volunteerEvents"."startTime" DESC 
        LIMIT 
          $1
      ) AS "Volunteer_volunteerEvents_m2m_1"
  ) AS "Volunteer_volunteerEvents_m2m" ON true 
WHERE 
  (
    "t1"."id" = $2 
    AND 1 = 1
  ) 
LIMIT 
  $3
```

**Reply by samuelcole** [reply]
sorry, just saw you all would prefer new issues for bug reports: https://github.com/prisma/prisma/issues/22311

**Reply by janpio** [reply]
All good, we put that comment in our backlog as well - but thanks for putting it into an issue. Now we are tracking that.

**Reply by janpio** [reply]
This has been fixed, and released with an earlier patch 5.7.1 and again with today's 5.8.0. 
Can you confirm this is fixed with these versions @samuelcole?

**Comment by BHesseldieck** [comment]
If I turn on the preview feature many of my queries fail with this error:

```js
"PrismaClientKnownRequestError: 
Invalid `.findUniqueOrThrow()` invocation in
/user/db/getUser.ts:20:6

  17   include: Prisma.Subset<T, Prisma.UserInclude>,
  18 ) {
  19   return db.user
→ 20     .findUniqueOrThrow(
        Inconsistent column data: Unexpected conversion failure from String(2023-06-09T23:40:31.6+00:00) to DateTime. 
        Reason: $trailing input at Object.handleRequestError (/web/node_modules/.pnpm/@prisma+client@5.7.0_prisma@5.7.0/node_modules/@prisma/client/runtime/library.js:125:6817)
    at Object.handleAndLogRequestError (/web/node_modules/.pnpm/@prisma+client@5.7.0_prisma@5.7.0/node_modules/@prisma/client/runtime/library.js:125:6151)
    at Object.request (/web/node_modules/.pnpm/@prisma+client@5.7.0_prisma@5.7.0/node_modules/@prisma/client/runtime/library.js:125:5859)
    at l (//web/node_modules/.pnpm/@prisma+client@5.7.0_prisma@5.7.0/node_modules/@prisma/client/runtime/library.js:130:10025)
    at <anonymous> (/web/node_modules/.pnpm/@prisma+client@5.7.0_prisma@5.7.0/node_modules/@prisma/client/runtime/library.js:34:550)"
```

if I turn it off it is working all fine again.
The column that is the problem is set like this: `createdAt    DateTime @default(now()) @map("created_at") @db.Timestamptz(6)`
The other column with the same time-value that could also cause it is: `updatedAt   DateTime @updatedAt @map("updated_at") @db.Timestamptz(6)`

query looks like this:
```
  db.user
    .findUniqueOrThrow({
      where: {
        id,
      },
      include: {
          userPermissions: {
              permission: true
          }
      }
    })
```

**Reply by janpio** [reply]
Thanks for the report!

I think that is known and tracked as https://github.com/prisma/prisma/issues/22293, work on fixing it is already in progress.

**Reply by janpio** [reply]
This should be fixed with an earlier patch release and also today's 5.8.0. Can you confirm @BHesseldieck?

**Comment by rcbevans** [comment]
Some feedback on the `relationJoins` preview feature.

I turned it on a few days ago and have had to just turn it off again because many of our list APIs that use skip/limit pagination with complex joins and filtering on join tables experienced an awful drop in performance with API calls usually taking ~0.2s suddenly taking 15+s and causing memory pressure on the backend.

From the logged queries from Prisma it appears the `relationJoins` version of the query was pulling every row in the database with all of the expensive and complex joined data and the rows were *then* being filtered through secondary queries.

**Reply by janpio** [reply]
That would definitely not be expected. Thanks for reporting.

Can you share an example of such a query, so we are sure we are looking at the right thing?
Can you potentially also share the SQL query that was executed?
Also your Prisma schema for context regarding indexes?
Thanks.

**Reply by BracketJohn** [reply]
Just to chime in here: We have experiences the same: Also a huge drop in performance when turning on `relationJoin` vs when turning it off (both while staying on v5.7.1).

**Reply by rcbevans** [reply]
I'll try share when I can.

**Reply by Weakky** [reply]
Hey @BracketJohn, similarly to @rcbevans, could you please share more information with us? Which database & version are you using, which queries are much slower, what's your Prisma Schema (even reduced to the minimum to reproduce the slow queries)?

Thank you!

**Reply by BracketJohn** [reply]
Hey @Weakky!

We are on v5.7.1 and using, here's out `client` section (I commented back in the `previewFeatures`, in our real app they are commented out, as it slows down the app so much):
```prisma
generator client {
  provider = "prisma-client-js"
  // TODO: Enabling this actually _reduces_ performance (albeit it should be increasing it). What we are experiencing is similar to the issue described here https://github.com/prisma/prisma/discussions/22288#discussioncomment-7847470
  previewFeatures = ["relationJoins", "nativeDistinct"]
}
```

-> I think someone else opened an issue that is maybe the same as this & they also added a schema to reproduce with: https://github.com/prisma/prisma/issues/22596

**Comment by andyjy** [comment]
`5.8.0` release notes documented `relationLoadStrategy: "query"` as "query — uses sub-queries to fetch relation data" - so I was surprised to see when I enabled `logLevel: ["query"]` that it actually reverted back to 2 separate, sequential queries as if `relationJoins` hadn't been enabled. 

I suspect this is a documentation terminology nit in the release notes (assuming the behaviour I'm seeing is expected) - but the mention of "sub-queries" led me to expect it would still use a single SQL request as per the typical usage of "subquery" to mean an inner or nested query (i.e. be some kind of alternative strategy for optimising joins, rather than effectively `relationJoinStrategy: "none"`)

**Reply by aqrln** [reply]
Hey @andyjy, good catch, and sorry for the confusion! I went ahead and fixed the wording in the release notes on GitHub.

`relationLoadStrategy: "query"` is indeed using separate queries to fetch relations, and is equivalent to the old behaviour without `relationJoins` preview feature. The idea is that you can opt-in to the old behaviour on per-query basis even when the new behaviour is enabled by default.

**Comment by andyjy** [comment]
(Cross-posting for visibility, hope helpful) - identified the cause of a bug causing incompatibility with driverAdapters causing fatal error when the included relation contains null fields: https://github.com/prisma/prisma/issues/22294#issuecomment-1889093869

**Reply by aqrln** [reply]
Sorry for not responding here quickly @andyjy, and thanks for helping out! We were aware of it (there are a few duplicate issues, including an internal one, so the context was lost and not easily available, sorry about this) but it's very good to have a confirmation. The fix is currently being worked on, but the very root cause was a little unexpected and ultimately involves an important third-party dependency that needs to be fixed upstream. We should hopefully be able to wrap it up before the release.

**Comment by renatocron** [comment]
Greetings Prisma Community,

I've been actively running a medium-scale project in a staging environment using the preview feature, which so far hasn't presented any issues. My views on topics like this issue https://github.com/prisma/prisma/issues/22479, which I believe reflects more of developer oversight than a genuine Prisma issue, are firm.

Initially, I held a subjective perspective on how Prisma developers envisioned the function's implementation, chiefly considering it from an infrastructural side. However, upon the release of version 5.8, my guidance on this has moderately shifted. Below, I detail this feature's implications on my project.

My project experienced considerable advantages on numerous endpoints when executing this featureFlag. This applies even when running in a virtually zero-latency network (API and database residing on a non-VM hardware, same machine, in docker). Although, in some situations, the tradeoff between several PostgreSQL JSONB serializations and Node.JS object decoding + instantiations did not performed better to a complete transition to "relationship" queries.

In that manner, considering marking bulky data tables in the schema (This could be as simple as a `@@heavy` or something like `@@external`) was a thought of mine around the 5.8 release, motivated by responses such as this one https://github.com/prisma/prisma/issues/22596 reporting adverse impacts upon enabling this feature

While we can suggest amendments to his code for superior query performance, remember that his query isn't wrong per se. Hence, as a framework, we hold the obligation to retain a decent level of performance for such queries. Also, this would help in paving a smoother path to preview this feature for architects (like marking the schema rather than modifying the query route, where the code resides in).

The position Prisma decides to adopt has a direct bearing on this matter. If we lean towards the view that "developers simply employ Prisma", it supports the notion that we should provide them with options that allow for behavior modifications independent of the code they're currently working with. Conversely, if we operate under the belief that "developers are fully aware of their needs", then it would logically lead us to consider making minor amendments to the API. However, my viewpoints have been nuanced due to a recently encountered post, I'm a bit lost as I'm no part of the prisma dev team, but if the later option is the one, I have a few ideias:

Given an opt-in for each query towards "get everything in two queries," (a minor rollback to the prior implementation where it always resolved in a single query for the joined field using some nifty parallel join :P). And like all magic, this is merely an illusion - developers must decide what should be joined inline (JSON with parallels) or queried independently, both happening within the same query.

So, for instance, if I have a user with multiple payments having various logs, each of which has a created_by User relation, a command such as the following could come in handy:

```js
const rows = await prisma.user.findFirst({
    data: {
        id: user.id,
    },
    select: {
        id: true, // get user id
        payment: {
            date: true, // get payment date,
            amount: true,            
            logs: {
               data: true,
               creator: {
                   select: {
                      id: true,
                      display_name:true
                   }
               }
            }
        },
    },
    prefetch: { 
        payment: {
            creator: false 
        }
    }
});
```

From Prisma's perspective, the optimization in Prisma ensures relations are fetched with improved efficiency while minimizing the need for multiple data roundtrips.


```sql
SELECT "id" 
FROM "User" 
WHERE "id" = :userId
LIMIT 1;
```
-- remember the LIMIT 1 , as this ia a findFirst, but this another topic, maybe

Then it would gather the payment data against the specific user and include their logs. Also, owing to the prefetch clause, Prisma would wisely omit the 'creator' data from the JOIN operation, generating a single inlined query:

```sql
SELECT "Payment"."date", "Payment"."amount", "Log"."data", "Log"."creator_id" 
FROM "Payment" 
LEFT JOIN "Log" ON "Payment"."id" = "Log"."payment_id"
WHERE "Payment"."user_id" = :userId;
```

Consequently, to retrieve the creator's information from the 'User' relation with the obtained creator_ids from the previous step, Prisma would generate:
```sql
SELECT "id", "display_name" 
FROM "User" 
WHERE "id" IN (:creatorIds);
```

These SQL queries reflect Prisma's optimization and ability to handle relational data fetching more effectively with custom control to the developer.


If you read until here, thank you.

**Reply by aqrln** [reply]
Thank you for the feedback @renatocron!

I don't know if you've seen it already, but we shipped a limited form of this in Prisma 5.8.0: you can specify whether to use JOINs or separate queries on a per Prisma query basis using the `relationLoadStrategy` argument, but it applies to all relations in the query.

Configuring it on a per-relation basis on any level of nested selections would be more powerful, and we want to have that as well, but it's not on the immediate plan right now. We'd like to consider implementing it once we are done with other improvements to `relationJoins` we are working on and have planned right now, so stay tuned! Also see my comment in response to @itsgiacoliketaco [below](https://github.com/prisma/prisma/discussions/22288#discussioncomment-8212858).

**Comment by itsgiacoliketaco** [comment]
This feature is most welcome, thank you!  The new `relationLoadStrategy: 'query' | 'join'` option is useful, but I'm wondering about the possibility of modifying the load strategy on a per-relation basis.  Consider:

```ts
db.posts.findMany({
  where: {
    // Only load posts for a specific author.
    authorId: '123',
  },
  include: {
    // Load `post.category` via JOIN.
    // Equivalent to `category: true` when `relationJoins` is enabled.
    category: { loadStrategy: 'join' },

    // Load `post.author` with a separate query.
    author: { loadStrategy: 'query' },
  },
})
```

Because of the `where` clause, we know that `post.author` will be the same for every post.  If there are many posts, it may be faster to load the `author` once in separate query instead of receiving N copies of the same author from the database.  However, `post.category` will differ between posts, so joins may be favorable for this relation.

I think this is similar to what @renatocron suggested above, but a different API.

**Reply by aqrln** [reply]
Thank you @itsgiacoliketaco! We did consider it, and you even guessed the exact API suggested in an internal document 😆
However, per-relation configuration is a lot trickier to implement than per-query configuration. For now we decided to implement the per-query strategy as it should already be very useful, was deliverable swiftly, and is forward-compatible with per-relation strategy.

Right now we are very busy working on higher priority improvements to `relationJoins` that will benefit the most users, but we can revisit this feature later! Could you please open a feature request as a GitHub issue to make it more visible, and perhaps gather interest from other users?

**Reply by itsgiacoliketaco** [reply]
Done!  https://github.com/prisma/prisma/issues/22759

Thanks for the transparency!  Seems like a good compromise.  Glad we're on the same page about the API when the per-relation strategy option is eventually introduced 😄

**Comment by kristianeboe** [comment]
Hey guys, I found a pretty concrete bug that unfortunately blocks our adoption of the relational joins. 

Basically, if you have a JSON[] field on a model, and you query that model directly, it resolves as an empty array or a populated array ( as it should). However, when you query that model through a join the empty Json[] fields resolve to null :O This breaks the typings, and the projects client side, because we often try to read .length of the presumed empty arrays.


Even more concrete example:

```
model Profile {
	id: Int
	userId: String
	links: Json[]
}

model User {
	id: String
	profile: Profile
}


```

Querying profile alone resolves correctly with links as [].

Querying User joined with Profile resolves with links to NULL (again, assuming they have not been populated).

Changing the relationLoadStrategy to "query" resolves the issue for now.


What do you guys think? Easy fix?

**Reply by janpio** [reply]
That sounds ... not great. Thanks for reporting!

Can you turn this into a full, standalone issue? We definitely want to look into that, and that is easier with a full issue with a filled issue template. Thank you!

**Reply by kristianeboe** [reply]
@janpio done :) https://github.com/prisma/prisma/issues/22923

**Comment by Sufiane** [comment]
Hey guys ! 

I'm currently trying this new feature (can't wait to have it in GA !), and my IDE throw me this error when i added the flag.

Is my IDE tripping or am I missing something ? I currently have the latest prisma release (v5.9)

<img width="1124" alt="Capture d’écran 2024-02-06 à 17 34 00" src="https://github.com/prisma/prisma/assets/10711632/37b11e74-f521-4199-8096-6a6438a55a2b">

**Reply by janpio** [reply]
Your IDE is running a Prisma extension for an older version of Prisma. If you are using VS Code, then https://marketplace.visualstudio.com/items?itemName=Prisma.prisma in its newest version definitely does support `relationJoins` there.

**Reply by Sufiane** [reply]
thanks for the quick reply @janpio ! I'm using we storm, I should be up-to-date. maybe an update is needed in the plugin?

**Reply by janpio** [reply]
https://plugins.jetbrains.com/plugin/20686-prisma-orm I assume then? I honestly do not know if and how IntelliJ updates the version of our "engine" in their plugin: https://github.com/JetBrains/intellij-plugins/tree/master/prisma That is a pretty good question.

Update: Hui, they are indeed pretty outdated: https://github.com/JetBrains/intellij-plugins/blob/a73c8ef111e2d7766098e5ede1723a90caf5f357/prisma/language-server/package.json#L6-L9 Seems in the past they did that via an issue in their issue tracker: https://youtrack.jetbrains.com/issue/WEB-63544

**Reply by Jolg42** [reply]
I'm not sure how their release process work but the update of the plugin was done in https://github.com/JetBrains/intellij-plugins/commit/eba9b8cfc9881033f47eeab639a6cf76c1221014 🎊

**Comment by michaelhays** [comment]
Very encouraging results with MySQL `relationJoins`! Some deeply nested queries going from ~2s to under 0.3s after enabling :partying_face:

I have encountered a few `Out of sort memory` errors, but that's due to the sheer size of our queries (6 levels of nesting, nearly 30 models in a single query). Our most gnarly query has Prisma generating an SQL query nearly 1,000,000 characters long :exploding_head:

I don't think there's anything to do on Prisma's end -- I'm planning to refactor these queries to optimize them.

Thanks Prisma team <3

**Comment by michaelhays** [comment]
Posting here for visibility -- I found a bug with the MySQL implementation that seems to convert decimals to floats when retrieving nested models: https://github.com/prisma/prisma/issues/23233

**Reply by michaelhays** [reply]
Fixed in 5.12.0! :tada:

**Comment by janpio** [comment]
3 new issues tracking the demand for the addition of this functionality for SQL Server, SQLite and MariaDB:
https://github.com/prisma/prisma/issues/23346, https://github.com/prisma/prisma/issues/23347, https://github.com/prisma/prisma/issues/23348

**Comment by jgnieuwhof** [comment]
Thanks for the feature!

We had it enabled for a little bit, but ran into an issue when using `@db.Oid` as a foreign key:

https://github.com/prisma/prisma/issues/23262#issuecomment-1998045146

**Comment by Scalahansolo** [comment]
@janpio I was hoping to opt into the `join` strategy on a per query basis as we test things out. But... from what I can tell there is no way to enable this feature, but set the global default to keep with the `query` strategy? 

There are a few queries in our setup that we want to be able to test but don't want to turn on this feature for all our queries right away.

**Comment by TheMC47** [comment]
Nice! We got quite a nice performance boost, but also found a small bug https://github.com/prisma/prisma/issues/24414

**Comment by iamconor** [comment]
Hi does the `relationJoins` flag get picked up by Prisma Accelerate? Or will Accelerate generate queries the 'old' Prisma way using several SELECTs and application-level joins?

**Reply by aqrln** [reply]
It will be used by Accelerate, yes.

**Reply by iamconor** [reply]
Thanks for the response @aqrln - just to confirm, when the `relationJoins` flag is on, is it currently used by Accelerate to generate db join queries?

**Reply by aqrln** [reply]
yes

**Reply by iamconor** [reply]
OK in that case we've been using `relationJoins` in production for >1 month now with no bugs. Thanks!

**Comment by key-moon** [comment]
Hey. Today, I encountered a bug that this feature flag seems to have caused.

Following a query with a schema like `A <-1--N-> B <-N--N-> C` generates an invalid SQL Query, and it causes a `The column t3.name does not exist in the current database.` error.
```ts
await prisma.a.findFirst({
  select: {
    bs: {
      where: {
        cs: {
          every: {
            name: "a"
          },
        },
      }
    }
  }
});
```

## Client Error
```
PrismaClientKnownRequestError: 
Invalid `prisma.a.findFirst()` invocation in
./script.ts:6:18

  3 const prisma = new PrismaClient()
  4 
  5 async function main() {
→ 6   await prisma.a.findFirst(
The column `t3.name` does not exist in the current database.
    at Ln.handleRequestError (./node_modules/.pnpm/@prisma+client@5.19.0_prisma@5.19.0/node_modules/@prisma/client/runtime/library.js:121:7753)
    at Ln.handleAndLogRequestError (./node_modules/.pnpm/@prisma+client@5.19.0_prisma@5.19.0/node_modules/@prisma/client/runtime/library.js:121:7061)
    at Ln.request (./node_modules/.pnpm/@prisma+client@5.19.0_prisma@5.19.0/node_modules/@prisma/client/runtime/library.js:121:6745)
    at async l (./node_modules/.pnpm/@prisma+client@5.19.0_prisma@5.19.0/node_modules/@prisma/client/runtime/library.js:130:9633) {
  code: 'P2022',
  clientVersion: '5.19.0',
  meta: { modelName: 'A', column: 't3.name' }
}
```

## DB Error
```
ERROR:  column t3.name does not exist at character 296
STATEMENT:  SELECT "t1"."id", "A_bs"."__prisma_data__" AS "bs" FROM "public"."A" AS "t1" LEFT JOIN LATERAL (SELECT COALESCE(JSONB_AGG("__prisma_data__"), '[]') AS "__prisma_data__" FROM (SELECT "t4"."__prisma_data__" FROM (SELECT JSONB_BUILD_OBJECT('id', "t3"."id", 'aId', "t3"."aId") AS "__prisma_data__", "t3"."name" FROM (SELECT "t2".* FROM "public"."B" AS "t2" WHERE "t1"."id" = "t2"."aId" /* root select */) AS "t3" /* inner select */) AS "t4" WHERE ("t4"."id") NOT IN (SELECT "t1"."A" FROM "public"."_BToC" AS "t1" INNER JOIN "public"."C" AS "j1" ON ("j1"."id") = ("t1"."B") WHERE ((NOT "j1"."name" = $1) AND "t1"."A" IS NOT NULL)) /* middle select */) AS "t5" /* outer select */) AS "A_bs" ON true LIMIT $2
```

I've created a PoC repository, so I would appreciate it if you could take a look.

https://github.com/key-moon/prisma-relationjoins-bug-reproduce

Thank you!

**Reply by aqrln** [reply]
Thanks, I opened an issue to track this: https://github.com/prisma/prisma/issues/25104

**Comment by casparolesen** [comment]
Really looking forward to using this in production! Any idea when it'll go into GA?

**Reply by aqrln** [reply]
Not in the next quarter (we do have something equally exciting planned though 😉) but hopefully we can prioritize this right after

**Reply by casparolesen** [reply]
Sounds good, thank you!

**Comment by wmadden** [comment]
A quick update, we plan to release this preview feature to GA between March - August 2025. Read more here: https://github.com/prisma/prisma/discussions/26136

**Comment by koko-fv** [comment]
bump, seems like a cool nice to have. Makes the generated queries more readable too

**Comment by fabregas4** [comment]
Any chance we could have a setting to specify the default strategy - query or join?

I want to have a ability to use 'join' in a few places without having to go and specify 'query' on every other single query in the app.

This would be useful for incremental adoption of 'join'. Otherwise we have to fully re-test every single part of the app once enabling the preview.

**Reply by renatocron** [reply]
I think it's doable by using $extension and some hacking on findMany/findFirst methods to autoInject if undef

**Reply by solokhind** [reply]
In my case I wanted to test it only in a mirror of my app instead of enabling it in all instances. I also like the idea of just toggling between the 2 via an env configuration. Here's how I achieved that:
```ts
.$extends({
    query: {
      $allOperations({ args, query }) {
        const relationLoadStrategy = process.env.PRISMA_RELATION_LOAD_STRATEGY;

        if (
          (relationLoadStrategy === "join" ||
            relationLoadStrategy === "query") &&
          !args?.relationLoadStrategy &&
          args &&
          // Only apply to operations that support relations (include/select)
          (args.include || args.select)
        ) {
          return query({
            ...args,
            relationLoadStrategy,
          });
        }

        return query(args);
      },
    },
  });
```

**Comment by AlexH6015** [comment]
I'm using prisma with mysql, after enabling the relationJoins flag, i started getting error on join tables that had bit type columns, they were configured as boolean in the prisma schema. i started getting error that string base64 "AA==" cannot be converted to boolean.

**Comment by Useems** [comment]
I would really like support for MariaDB. I don't want to have to migrate my production database to MySQL 5.7 and then to 8 just because of Prisma.

**Comment by DerTimonius** [comment]
any updates on the ETA of the general availability of `relationJoins`? the initial post suggests that it will be available by the end of August 2025
